### PR TITLE
#174838707 A user should be able to reset their password [WIP]

### PR DESCRIPTION
**Pull request by balldash**
### What does this PR do?
Allow users to reset their passwords

### Description of the Task to be completed?
- create a recover password route
- create a reset password route
- create a recover password controller
- create a reset password controller
- configure nodemailer to send emails
- configure handbars.js for email template
- Sign up for sendgrid email service
- Write tests for the controllers and routes

### How should this be manually tested?
1. Clone the PR
2. Checkout to the branch ```ft-user-can-reset-password-174838707```
3. Install dependencies using ```npm install```
4. Create a ```.env file``` and define the following:
   * ```DEV_DATABASE_URL = <development database url>```
   * ```TEST_DATABASE_URL = <test database url>```
   * ```DATABASE_URL = <production database url>```
   * ```SENDGRID_API_KEY= <your send grid api key>```
   * ```SENDGRID_EMAIL = <your SENDGRID EMAIL goes here>```
5. Run ```npm run migrate``` to create the database tables
6. Run ```npm run seed``` to seed data to the database
7. Run ```npm run start-dev``` to start the server
8. Go to [Temp Mail](https://temp-mail.org/en/) to get a temporally email for testing
9. Make a POST request to the end point: (http://localhost:3000/api/v1/users/signup) using postman with the following object in the body: ```{ "username": <choose a username>, "email": <the temporal email from temp-mail>, "password": "123456" } ```
10. Make a POST request to the end point: (http://localhost:3000/api/v1/users/recover) with the following object in the body: ```{ "email": <the temporal email from temp-mail> }```
11. Check the inbox on temp-mail for reset password mail
12. Copy the link in the email and use it in postman to make a POST request with the following object in the body: ```{ "newPassword": <your new password>}```

### Link to architecture
[Flowchart](https://docs.google.com/document/d/1dPpdT_tc1y_OmolHHXZf-ghmF554ctTYL01Bg6LTe7k/edit?usp=sharing)

### What are the relevant pivotal tracker stories?

[#174838707](https://www.pivotaltracker.com/story/show/174838707)

### CODE REVIEW

File: src/controllers/resetPasswordController.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+          const hashedPass = hashPassword(newPassword);
+          try {
+            db.Users.update({ password: hashedPass, }, {
+              where: { id: user.id },
+              returning: true,
+              plain: true
+            });
+          } catch (error) { throw error; }
+          return res.status(200).json({ status: 200, success: "password has been reset" });
+        } catch (error) {
```
Comment (Extremophile-ai) [inline src/controllers/resetPasswordController.js:77]: I think the status code here is wrong.
A more appropriate statusCode should be 500 for server error or 409 for conflict

**Review by Extremophile-ai — CHANGES_REQUESTED** [review]

File: src/controllers/resetPasswordController.js (hunk 1 above)
Comment (Extremophile-ai) [inline src/controllers/resetPasswordController.js:77]: I think the status code here is wrong.
A more appropriate statusCode should be 500 for server error or 409 for conflict

**Review by Extremophile-ai — CHANGES_REQUESTED** [review]
I think the status code here is wrong.
A more appropriate statusCode should be 500 for server error or 409 for conflict

**Review by Extremophile-ai — CHANGES_REQUESTED** [review]
Your use of status code in your resetPasswordController is wrong.

Kindly research on statuscodes and use as appropriate

**Review by Extremophile-ai — DISMISSED** [review]
Requested changes has been made.

But one of your test is failing.

**Review by Fiyin-Anne — APPROVED** [review]

**Review by bellogo — APPROVED** [review]

File: src/tests/controllers/resetPasswordTest/reset-test-data.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    password: "thunderbolt",
+  })
+};
+const notUser = {
+  email: "jadeclaw@notuser.com",
+};
+
+const newPassword = {
+  newPassword: "firebolt",
+  // toJSON: () => ({
```
Comment (codeliezel) [inline src/tests/controllers/resetPasswordTest/reset-test-data.js:21]: remove this commented code

File: src/controllers/resetPasswordController.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
+import db from "../models/index";
+import signToken from "../utilities/signToken";
+import hashPassword from "../utilities/hashPassword";
+import sendGrid from "../utilities/sendgrid";
+
+require("dotenv").config();
+const jwt = require("jsonwebtoken");
+
+let hostURL;
```
Comment (codeliezel) [inline src/controllers/resetPasswordController.js:9]: you are not using this, you can remove it `let hostURL;`

**Review by codeliezel — APPROVED** [review]

**Pull request merged**